src/lib/operators/join_hash.cpp

```diff
@@ -698,55 +723,91 @@ class JoinHash::JoinHashImpl : public AbstractJoinOperatorImpl {
+      auto pos_list_ptrs = std::make_shared<PosLists>(input_table->chunk_count());
+      auto pos_lists_iter = pos_list_ptrs->begin();
+
+      for (ChunkID chunk_id{0}; chunk_id < input_table->chunk_count(); chunk_id++) {
+        // This works because we assume that the columns have to be either all ReferenceColumns or none.
+        const auto& ref_column =
+            *static_cast<const ReferenceColumn*>(input_chunks[chunk_id]->columns()[column_id].get());
+        *pos_lists_iter = ref_column.pos_list().get();
+        ++pos_lists_iter;
+      }
+
+      auto iter = shared_pos_lists_by_pos_lists.emplace(*pos_list_ptrs, pos_list_ptrs).first;
+
+      *pos_lists_by_column_it = iter->second;
+      ++pos_lists_by_column_it;
+    }
+
+    return pos_lists_by_column;
+  }
+
   static void write_output_chunks(const std::shared_ptr<Chunk>& output_chunk,
-                                  const std::shared_ptr<const Table> input_table, PosList& pos_list,
+                                  const std::shared_ptr<const Table> input_table,
+                                  const PosListsByColumn& input_pos_list_ptrs_sptrs_by_column, PosList& pos_list,
                                   bool is_ref_column) {
     if (pos_list.empty()) return;
 
+    std::map<PosListsSPtr, std::shared_ptr<PosList>> output_pos_list_by_input_pos_list_ptrs_sptr;
+
     // Add columns from input table to output chunk
     for (ColumnID column_id{0}; column_id < input_table->column_count(); ++column_id) {
       std::shared_ptr<BaseColumn> column;
 
       if (is_ref_column) {
-        auto ref_col =
-            std::dynamic_pointer_cast<const ReferenceColumn>(input_table->get_chunk(ChunkID{0})->get_column(column_id));
-
-        // Get all the input pos lists so that we only have to pointer cast the columns once
-        auto input_pos_lists = std::vector<std::shared_ptr<const PosList>>();
-        for (ChunkID chunk_id{0}; chunk_id < input_table->chunk_count(); chunk_id++) {
-          // This works because we assume that the columns have to be either all ReferenceColumns or none.
-          auto ref_column =
-              std::dynamic_pointer_cast<const ReferenceColumn>(input_table->get_chunk(chunk_id)->get_column(column_id));
-          input_pos_lists.push_back(ref_column->pos_list());
-        }
-
-        // Get the row ids that are referenced
-        auto new_pos_list = std::make_shared<PosList>();
-        for (const auto& row : pos_list) {
-          if (row.chunk_offset == INVALID_CHUNK_OFFSET) {
-            new_pos_list->push_back(row);
-          } else {
-            new_pos_list->push_back(input_pos_lists.at(row.chunk_id)->at(row.chunk_offset));
+        const auto& input_table_pos_lists = input_pos_list_ptrs_sptrs_by_column[column_id];
+
+        auto iter = output_pos_list_by_input_pos_list_ptrs_sptr.find(input_table_pos_lists);
+        if (iter == output_pos_list_by_input_pos_list_ptrs_sptr.end()) {
```

Comment:
Naming: I would call it `output_pos_list_cache` or something like that.

Reply:
jap